// ngx-otp-input/esm2022/lib/directives/ariaLabels.directive.mjs
import { Directive, Input, ContentChildren, } from '@angular/core';
import * as i0 from "@angular/core";
export class AriaLabelsDirective {
    ngAfterContentInit() {
        this.setAriaLabelsAttrs();
    }
    getDefaultAriaLabelText(index) {
        return `One Time Password Input Number ${index + 1}`;
    }
    setAriaLabelsAttrs() {
        this.inputs.forEach((input, index) => {
            input.nativeElement.setAttribute('aria-label', this.ngxOtpAriaLabels[index] ?? this.getDefaultAriaLabelText(index));
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AriaLabelsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.1", type: AriaLabelsDirective, isStandalone: true, selector: "[ngxOtpAriaLabels]", inputs: { ngxOtpAriaLabels: "ngxOtpAriaLabels" }, queries: [{ propertyName: "inputs", predicate: ["otpInputElement"], descendants: true }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AriaLabelsDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngxOtpAriaLabels]',
                }]
        }], propDecorators: { inputs: [{
                type: ContentChildren,
                args: ['otpInputElement', { descendants: true }]
            }], ngxOtpAriaLabels: [{
                type: Input
            }] } });
