// ngx-otp-input/esm2022/lib/directives/autoFocus.directive.mjs
import { ContentChild, Directive, Input, } from '@angular/core';
import * as i0 from "@angular/core";
export class AutoFocusDirective {
    ngAfterContentInit() {
        if (this.ngxAutoFocus && this.firstInput) {
            this.firstInput.nativeElement.focus();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AutoFocusDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.1", type: AutoFocusDirective, isStandalone: true, selector: "[ngxAutoFocus]", inputs: { ngxAutoFocus: "ngxAutoFocus" }, queries: [{ propertyName: "firstInput", first: true, predicate: ["otpInputElement"], descendants: true }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AutoFocusDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngxAutoFocus]',
                }]
        }], propDecorators: { firstInput: [{
                type: ContentChild,
                args: ['otpInputElement', { static: false }]
            }], ngxAutoFocus: [{
                type: Input
            }] } });
