// ngx-otp-input/esm2022/lib/directives/autoBlur.directive.mjs
import { ContentChildren, Directive, Input, } from '@angular/core';
import * as i0 from "@angular/core";
export class AutoBlurDirective {
    constructor() {
        this.inputHTMLElements = [];
    }
    ngOnChanges(changes) {
        if (this.ngxAutoBlur &&
            this.inputHTMLElements.length > 0 &&
            changes['isFormValid'].currentValue) {
            this.inputHTMLElements.forEach((input) => {
                input.blur();
            });
        }
    }
    ngAfterContentInit() {
        this.inputs.forEach((input) => {
            this.inputHTMLElements.push(input.nativeElement);
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AutoBlurDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.1", type: AutoBlurDirective, isStandalone: true, selector: "[ngxAutoBlur]", inputs: { ngxAutoBlur: "ngxAutoBlur", isFormValid: "isFormValid" }, queries: [{ propertyName: "inputs", predicate: ["otpInputElement"], descendants: true }], usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: AutoBlurDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngxAutoBlur]',
                }]
        }], propDecorators: { inputs: [{
                type: ContentChildren,
                args: ['otpInputElement', { descendants: true }]
            }], ngxAutoBlur: [{
                type: Input
            }], isFormValid: [{
                type: Input
            }] } });
