// ngx-otp-input/esm2022/lib/directives/inputNavigations.directive.mjs
import { ContentChildren, Directive, EventEmitter, HostListener, Input, Output, } from '@angular/core';
import * as i0 from "@angular/core";
export class InputNavigationsDirective {
    constructor() {
        this.inputsArray = [];
        this.valueChange = new EventEmitter();
    }
    ngAfterContentInit() {
        this.inputsArray = this.inputs.toArray();
    }
    findInputIndex(target) {
        return this.inputsArray.findIndex((input) => input.nativeElement === target);
    }
    setFocus(index) {
        if (index >= 0 && index < this.inputs.length) {
            this.inputsArray[index].nativeElement.focus();
        }
    }
    onArrowLeft(event) {
        const index = this.findInputIndex(event.target);
        if (index > 0) {
            this.setFocus(index - 1);
        }
    }
    onArrowRight(event) {
        const index = this.findInputIndex(event.target);
        if (index < this.inputs.length - 1) {
            this.setFocus(index + 1);
        }
    }
    onBackspace(event) {
        const index = this.findInputIndex(event.target);
        if (index >= 0) {
            this.valueChange.emit([index, '']);
            this.setFocus(index - 1);
            event.preventDefault();
        }
    }
    onKeyUp(event) {
        const index = this.findInputIndex(event.target);
        if (event.target.value?.match(this.regexp)) {
            this.valueChange.emit([index, event.target.value]);
            this.setFocus(index + 1);
        }
        else {
            this.inputsArray[index].nativeElement.value = '';
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: InputNavigationsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.1", type: InputNavigationsDirective, isStandalone: true, selector: "[ngxInputNavigations]", inputs: { regexp: "regexp" }, outputs: { valueChange: "valueChange" }, host: { listeners: { "keydown.arrowLeft": "onArrowLeft($event)", "keydown.arrowRight": "onArrowRight($event)", "keydown.backspace": "onBackspace($event)", "input": "onKeyUp($event)" } }, queries: [{ propertyName: "inputs", predicate: ["otpInputElement"], descendants: true }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: InputNavigationsDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngxInputNavigations]',
                }]
        }], propDecorators: { inputs: [{
                type: ContentChildren,
                args: ['otpInputElement', { descendants: true }]
            }], regexp: [{
                type: Input
            }], valueChange: [{
                type: Output
            }], onArrowLeft: [{
                type: HostListener,
                args: ['keydown.arrowLeft', ['$event']]
            }], onArrowRight: [{
                type: HostListener,
                args: ['keydown.arrowRight', ['$event']]
            }], onBackspace: [{
                type: HostListener,
                args: ['keydown.backspace', ['$event']]
            }], onKeyUp: [{
                type: HostListener,
                args: ['input', ['$event']]
            }] } });
