// ngx-otp-input/esm2022/lib/directives/paste.directive.mjs
import { ContentChildren, Directive, EventEmitter, HostListener, Input, Output, } from '@angular/core';
import * as i0 from "@angular/core";
export class PasteDirective {
    constructor() {
        this.handlePaste = new EventEmitter();
    }
    onPaste(event) {
        event.preventDefault();
        const clipboardData = event.clipboardData?.getData('text');
        if (clipboardData && this.regexp.test(clipboardData)) {
            const values = clipboardData.split('');
            this.inputs.forEach((input, index) => {
                if (values[index]) {
                    input.nativeElement.value = values[index];
                }
            });
            this.handlePaste.emit(values);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: PasteDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.1", type: PasteDirective, isStandalone: true, selector: "[ngxOtpPaste]", inputs: { regexp: "regexp" }, outputs: { handlePaste: "handlePaste" }, host: { listeners: { "paste": "onPaste($event)" } }, queries: [{ propertyName: "inputs", predicate: ["otpInputElement"], descendants: true }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: PasteDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngxOtpPaste]',
                }]
        }], propDecorators: { inputs: [{
                type: ContentChildren,
                args: ['otpInputElement', { descendants: true }]
            }], regexp: [{
                type: Input
            }], handlePaste: [{
                type: Output
            }], onPaste: [{
                type: HostListener,
                args: ['paste', ['$event']]
            }] } });
